<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-A01793654/blob/main/Actividad7_IBM(Mod_4_MD).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ciencia y analítica de datos
## Actividad 7
### Cursos de IBM
### Data Analysis With Python
### Modulo 4
##### Profesor: Jobish Vallikavungal Devassia
##### Profesor Tutor: Roberto Antonio Guevara González
##### Bernardo Mijangos Flores
##### Matricula A01793654
### Laboratorio 4

# Desarrollo del modelo

Objetivos:

Desarrollar modelos de predicción 

En esta sección, desarrollaremos varios modelos que predecirán el precio del automóvil utilizando las variables o características. Esto es solo una estimación, pero debería darnos una idea objetiva de cuánto debería costar el automóvil.

En el análisis de datos, a menudo usamos el desarrollo de modelos para ayudarnos a predecir futuras observaciones a partir de los datos que tenemos.

Un modelo nos ayudará a comprender la relación exacta entre diferentes variables y cómo se utilizan estas variables para predecir el resultado.

In [6]:
# Importamos la librerias

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.linear_model import LinearRegression   # para la regresion lineal

In [4]:
# obtenemos los datos del ejercicio del siguiente link

path = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/automobileEDA.csv'

In [5]:
# obtenemos el dataframe con los datos

df = pd.read_csv(path)
df.head()

symboling  normalized-losses         make aspiration num-of-doors  \
0          3                122  alfa-romero        std          two   
1          3                122  alfa-romero        std          two   
2          1                122  alfa-romero        std          two   
3          2                164         audi        std         four   
4          2                164         audi        std         four   

    body-style drive-wheels engine-location  wheel-base    length  ...  \
0  convertible          rwd           front        88.6  0.811148  ...   
1  convertible          rwd           front        88.6  0.811148  ...   
2    hatchback          rwd           front        94.5  0.822681  ...   
3        sedan          fwd           front        99.8  0.848630  ...   
4        sedan          4wd           front        99.4  0.848630  ...   

   compression-ratio  horsepower  peak-rpm city-mpg highway-mpg    price  \
0                9.0       111.0    5000.0       21          27  13495.0   
1                9.0       111.0    5000.0       21          27  16500.0   
2                9.0       154.0    5000.0       19          26  16500.0   
3               10.0       102.0    5500.0       24          30  13950.0   
4                8.0       115.0    5500.0       18          22  17450.0   

  city-L/100km  horsepower-binned  diesel  gas  
0    11.190476             Medium       0    1  
1    11.190476             Medium       0    1  
2    12.368421             Medium       0    1  
3     9.791667             Medium       0    1  
4    13.055556             Medium       0    1  

[5 rows x 29 columns]

# 1. Regresión lineal y regresión lineal múltiple

## Regresión lineal simple

La regresión lineal simple es un método que nos ayuda a comprender la relación entre dos variables:

*   El predictor/variable independiente (X)
*   La respuesta/variable dependiente (que queremos predecir)(Y)

El resultado de la regresión lineal es una función lineal que predice la variable de respuesta (dependiente) en función de la variable predictora (independiente).

$$
Y: Response \ Variable\\\\\\\\\\
X: Predictor \ Variables
$$

<b>Función lineal</b>
$$
Yhat = a + b  X
$$

*   se refiere a la intersección de la línea de regresión, en otras palabras: el valor de Y cuando X es 0
*   se refiere a la pendiente de la línea de regresión, en otras palabras: el valor con el que cambia Y cuando X aumenta en 1 unidad


In [9]:
# Creando la regresion lineal

lm = LinearRegression()
lm

LinearRegression()

Se requiere ver cómo highway-mpg puede ayudarnos a predecir el precio del automóvil. Usando una regresión lineal simple, Se crea una función lineal con "highway-mpg" como la variable predictora y el "price" como la variable de respuesta.

In [11]:
# asignamos la variable predictora y de respuesta en X y Y

X = df[['highway-mpg']]
Y = df['price']

In [12]:
# Ajuste el modelo lineal

lm.fit(X,Y)

LinearRegression()

In [13]:
# Generamos una prediccion

Yhat=lm.predict(X)
Yhat[0:5]   

array([16236.50464347, 16236.50464347, 17058.23802179, 13771.3045085 ,
       20345.17153508])

In [14]:
# Valor del intercepto

lm.intercept_

38423.3058581574

In [15]:
# Valor de la pendiente

lm.coef_

array([-821.73337832])

Con los valores reales obtenemos:

Precio = 38423.31 - 821.73 x highway-mpg

# Pregunta 1a)

Cree un objeto de regresión lineal llamado "lm1".

In [16]:
lm1 = LinearRegression()
lm1

LinearRegression()

# Pregunta  1b):
¿Entrenar el modelo utilizando "engine-size" como variable independiente y "price" como variable dependiente?

In [17]:
lm1.fit(df[['engine-size']], df[['price']])
lm1

LinearRegression()

# Pregunta 1c):

Encuentra la pendiente y el intercepto del modelo.

In [21]:
print('Pendiente',lm1.coef_)
print('Intercepto',lm1.intercept_)

Pendiente [[166.86001569]]
Intercepto [-7963.33890628]


# Pregunta 1d):

¿Cuál es la ecuación de la línea predicha? Puede usar x y yhat o  "engine-size" o "price".

In [26]:
Yhat=-7963.34 + 166.86*X

In [27]:
Price=-7963.34 + 166.86*df['engine-size']